### ISSUE TITLE
`Tuple & nominal` not recognised as tuple in spread parameter

### ISSUE THREAD

**Issue by Dionysusnu** [microsoft/TypeScript#45371]
# Bug Report

### 🔎 Search Terms

Nominal, Spread, Tuple, 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about spread and intersections

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEEQBd5QFzwNoDsCuAtgEYgwA08+xpFVJMAuvAGTwDe8A+jgPYECWOKBE4ZKIAG6l4AXwCwAKEUAzPDjDJ+PHPCIAKAHRGADrCgEAzplyF6tWzUoPGASnaLF8pQv1GDUF0UgA)

### 💻 Code

```ts
declare let a: [number, number, number] & { _nominal_a: never }
function b(...params: [number, number, number]) {}
b(...a) // Error: A spread argument must either have a tuple type or be passed to a rest parameter.
```

### 🙁 Actual behavior

The error in the code example is thrown, because TypeScript does not detect it as being a tuple.
In roblox-ts, we use a nominal marker type to indicate a special kind of tuple. This bug means that this tuple type can then not be spreaded into a fixed argument function. Our issue: https://github.com/roblox-ts/roblox-ts/issues/1537
### 🙂 Expected behavior

This is a tuple type, and should therefore be recognised as such.

**Comment by S0AndS0** [comment]
I recently ran into this error too, and worked around it by _hinting_ the type via the `as` keyword, eg.

```typescript
function mergeDictionaries(...dictionaries: { [key: string]: unknown }[]): {
  [key: string]: unknown;
} {
  return Object.assign(
    ...(dictionaries.map((dictionary) => {
      return JSON.parse(JSON.stringify(dictionary));
    }) as [{ [key: string]: unknown }])
  );
}
```

... without the `[{ [key: string]: unknown }])` bit, an error similar to the following pops;

>     <file>.ts|<C> col <L> error| A spread argument must either have a tuple type or be passed to a rest parameter.

What I find somewhat trubling, beyond TypeScript _loosing the plot_ in the first place, is that `Array<T>` will produce the same error, eg.

```typescript
function mergeDictionaries(...dictionaries: { [key: string]: unknown }[]): {
  [key: string]: unknown;
} {
  return Object.assign(
    ...(dictionaries.map((dictionary) => { /* ← Error on this line */
      return JSON.parse(JSON.stringify(dictionary));
    }) as Array<{ [key: string]: unknown }>) /* ← Why is [{}] required? */
  );
}
```

**Comment by Dionysusnu** [comment]
> What I find somewhat trubling, beyond TypeScript _loosing the plot_ in the first place, is that `Array<T>` will produce the same error, eg.

This usually makes sense, because you don't know the Array's length.
I believe this not working in this case is a separate missing behaviour, which is that normal parameters need to be filled manually, not being considered to be part of the rest parameter. What I mean by that is that a function like `a(obj: object, ...other: object[])` will not allow a call like `a(...objects)` where `objects` is `Array<object>`, despite this being type-compatible.

**Comment by S0AndS0** [comment]
Hmm, I supose I'll have to dig deeper into how TypeScript parses things. Because I'm still a bit befuddled that `as Array<{ [key: string]: unknown }>` throws errors but using the same hint for a parameter doesn't present issues, eg.

```typescript
function mergeDictionaries(
  ...dictionaries: Array<{ [key: string]: unknown }>  /* ← Totally okay */
  // ...dictionaries: [{ [key: string]: unknown }]    /* ← this is fine too */
): {
  [key: string]: unknown;
} {
  return Object.assign(
    ...(dictionaries.map((dictionary) => {
      return JSON.parse(JSON.stringify(dictionary));
    }) as [{ [key: string]: unknown }])
    // }) as { [key: string]: unknown }[])       /* ← Nope */
    // }) as Array<{ [key: string]: unknown }>)  /* ← no joy either */
  );
}
```

**Comment by darklight9811** [comment]
I'm mapping a dispatch into a object and this is freaking my mind.

``` typescript
// not actual example but close
Object.fromEntries(
  Object.entries(actions)
    .map(entry => [
      entry[0],
      (...args: Parameters<typeof entry[1]>) => dispatch(entry[1](...args))
    ])
)
```

There is no reason for that be a problem, mainly because not allowing arrays as spread arguments should be a linter problem, not a compiler problem.

**Comment by Dionysusnu** [comment]
Made a [Playground Link](https://www.typescriptlang.org/play?target=6#code/MYewdgzgLgBAhsKBLcEBcMBKBTUAnAEwB5o8kwBzAGhgAoBKGAXgD4ZTyK2mYBvAXwCwAKBEFcAGzh5sMUJFgEkEAA5wowABYZa0ihg6VGrGADcQSAgG4RIgPIAjAFa4oAOgBmeEAFsAomBQZNgQtCIwMI4uiG7YgcGhCMio9OERMG4+cCq0cUEAnsxsANpp6TB5ePnFAAwAulRl6bRurXroMAAK0nA+2FDYeBBEUPkq2CAeFfHVAIx1LMZsSqrqWrkzxfMtbXgUEPSpwuV1R-RAA) for your code example, and I'm failing to spot any issue in this. In any case, it's probably similar to the above comment by `S0AndS0`, and as explained not a related problem.

**Comment by scottwillmoore** [comment]
I am not sure if this is related, or supposed to occur... However, when attempting to brand or flavor a tuple it appears that TypeScript is no longer able to restrict property access past the length of the tuple. 

**Example**

```ts
declare const id: unique symbol;
type Id<T> = { readonly [id]: T };

type Brand<T, U> = T & Id<U>;
type PartialBrand<T, U> = T & Partial<Id<U>>; // Also known as Flavor.

declare const pointId: unique symbol;
type Point = PartialBrand<[number, number], typeof pointId>;

const point = (x: number = 0, y: number = 0): Point => [x, y];

const add = (a: Point, b: Point, c: Point = point()): Point => {
  c[0] = a[0] + b[0];
  c[1] = a[1] + b[1];
  return c;
};

const a = point(0, 1);
const b = point(2, 3);
const c = add(a, b);

console.log(c);

c[0] = 4;
c[1] = 5;
c[2] = 6; // Should be an error!

console.log(c);
```

[Playground Link](https://www.typescriptlang.org/play?target=99&jsx=0#code/CYUwxgNghgTiAEYD2A7AzgF3gS2ALngFcVsBHQhNATwFsAjJCAbgFgAoDKgBwQElgAPABUAfPAC88AN7w4UYKghV4AbVwBdAkPgBfVm3ace8AEIwoKQUIA08AKpjJ2gGTx+Ah-qMIACrAzYUBBmFla2DhLwLvB+MAFBAu4OIkzwAPRp8ACCEGhI8ADWKEgA7ijwUGjwAGLQAG5IMAB07OygkLAIyOhYXEjYKBj8BMRkFPDU9Ixe3L79g5Gx8cHmlgIqKIT0IDC2m9sw6rbeSABm8H0DQ8AprWzdmBfzWJIAFAAeBPt0O5EADLYqF8tj8YP8AJQEHzPCRiFTvQHqfTsB5YeTASKvKBQ562Og4q62MAEhaSS6DV7gyExGHiMRSdjwJmIFR-dSRKCs9kAangdC5+mZLIAjOzJJzRfBefzRYLmXAMIQYOUwPo9HdURVIuSMK8AfBheD9Jq6Nrnq8AEy2ADMRpRqEeYA5wGAWLxdoM9wdjBATQgSAA5q8wB6UVzIgAWY0qSWSACs0YtYvgADZUhl4ABlAAWSEIEAxPwq5R2MEaAEINd6IL7-UGQ0wgA)

It appears that the expansion of the type includes `[x: number]: number`. Maybe I should open this a separate issue?

```ts
type DeepExpand<T> = T extends object ? T extends infer O ? { [K in keyof O]: DeepExpand<O[K]> } : never : T;
type ExpandPoint = DeepExpand<Point>;

// type ExpandPoint = {
//   [x: number]: number;
//   ...
// }
```

**Comment by TomRaaff** [comment]
Here is another instance of (probably) the same error: 
[Playground link](https://www.typescriptlang.org/play?ts=5.2.0-beta#code/PTAEBUAsFNQIwK4HNQEsDOaB2oAuNQBlaAJwDdUBjaAJWiQ1xIE8A6XTAEwHsB3HONAA2fVgCgQoGtwS5oxclWjtMAQyydQAGW6VVuVNywKK1FaFUlYRoc1BWAjglRXNAM24lQlI3IAeuOISYNKy8qSmyhxi0H4ADp643kKq6JihciZKoADeAL5iMfGJoG4IWJQGRvYymRFKAELOQpykABQAlLlioPbQuAgkOFjQvFK14YrUnQDcYgWFi4WSOnpVxvVm0bEJJEmUKWnauvqGG1OwOT3eRuhMCJWebXEkqGT6sFaqPFi2NWFZagALnGAM20C6+XmhR2JTKFXWoBEazOgOgTVQLXaJAmaJBGUmkUh1ysAyGoBGY1WpyMaLaOLBFw6cwWS2WYDRdAYdxYKjEuGYcVgXMYLFAAF5ur0GXULiCrr1eqgsHd1MDQbLIqAAD6gcqtNzK6CcOaK+DNVokEECoXcNz-TWNC2kU2gPKu5E086ReXXJUq3Bq6Ag6nrNE6vUaaCGkYmv3mzGW62C6B2pEnMPgjFYkiu93zOb8lNEcEAOVUAFtYJKANbQZhpkU85hzG2wACSAaDAB5CKBYnINJg0eWqwA+CWgUtGUsIIQpOBCaDdptMZgAbUIAF11wAiZWqirQXdbset4vZy0AMSwvf7AWgQ5LF1H0AnktXLE3O93iETpBPQsfADPpuTXfF6FFOxJQVB1CSUX0zQPQMjxBfVoyNTgABp4z-HMQRleDqEvUgcN6PIyPTFFaXBRDFWQoM0KjGNjUo3o8KTKivTREiSEoiiCxhYo9lKcpKjObwvjkO8B0fThhzLSs3zaa4sCUkFCEo1ohSjCpUGgdAQQABUsJS5BIdBu14m9ezHMccI6EFO0PahbKlG4QMfNdJysMCvzUqst1dVB7TaABCLzeQYo9iTNYD0G4JdWBEJA2m0uTH0ofT0GZeNJCgUhYHQaBoArTBcG4eBYFUc0UGVCBCFYUAAHJ0t0rKDOatBMB8EgrEqP5lTcUhXAsTA2nXLcI3XQi8Q1IjoC3Do2NASQAE0ZG8dRQGK2B8H0eA7EgbgyFIZUUBO0g8AIZr4sS5QUrS6AdNaPSDI6LrVTkKssFwV1ekkJpcCw0pPB27gqz6VIjBB8A+x4AysGapJ+u4JAsAwWBUCSVILDwBA4iXEHeEgKhID6dA5w4bALBwOGACYAFZGYANn7PrPHjSK2Gi6hJ251gOPaVgRba16OpyvMSX6QYcAF3noBZIA)

A TS2556 (A spread argument must either have a tuple type or be passed to a rest parameter.) error is thrown whenever I try to spread the contents of a tuple into another function. _Even though that tuple is inferred as the correct type._

I have tested this on all versions available in the playground.

**Comment by RA80533** [comment]
> It appears that the expansion of the type includes `[x: number]: number`. Maybe I should open this a separate issue?

Good observation. Once the index signature is removed, the tuple behaves as expected.

```ts
// https://stackoverflow.com/a/66252656
type RemoveIndex<T> = {
  [ K in keyof T as
    string extends K
    ? never
    : number extends K
    ? never
    : symbol extends K
    ? never
    : K
  ] : T[K];
};

// -snip-
type PartialBrand<T, U> = RemoveIndex<T> & Partial<Id<U>>; // Also known as Flavor.
//                        ^^^^^^^^^^^
```

[Playground link](https://www.typescriptlang.org/play?target=99&jsx=0&ssl=20&ssc=1&pln=18&pc=1#code/PTAEAsBdIBwZwFwjpAhgYwNYHsBuBTAJwDMAbbAdwDp1sBbYVYANmYCYBWN5j5gWABQkAJ4x8oAEr46efAEkAdgBN8ADwA8AFQB8oALygA3oNCgA2qADSoAJYLQmfMOzFQm0KjgnToFITsA5qBqkPjKcFbepgD8oAr4BIRRoAhxAK50AEZEwaqh4ZECPqCx8YnJqXDCWdikuflKEZbJpQlEFYWmALopbmaWXQDcggC+wwKCKuikqITitAootkqpaQo2AI5p4lU1pOMiYqBySlq6Boagc6hK2AqkwuY2Sl2p7mOCgofiAEKEqMotAAaUAAVXOblAADJjqdwQdROIAAqzSA2VCkP4A06aEHg-SSaSyRQqDQ6aGgFGENEY9QndTg7SDUAgUAAQVIcGwDgUlHsnlAADEZrhsIQqIJWcVpTLZaAAHqKpXK+WfARTGZzUALJYwbB2SAnVbrLY7aqZWoIo5I-UKSAEqk0zH-QFmBQZbKEEHurJELog74uUB6g0nJlqnX2kN2gkAClUqR9noJAAYQcJEx6cgYUwBKVI2g36XRmVTpoYRu5LG5KOOoAu2yAgzINg0g9CtmMGaOQWO5-OUxvFozJdBmFM9Ayocc9ADUoEyM-GxTHAEZJx4zOvQPPF+vlz45pA0oR7Ohxh8JgJIx4CT3Y2nQKvc+Mb5k743Y2wQQBmF+CG90AJGtY1QZt-yvHVanwKhyACWN0AggCZwJAAWV8tw3DgMLYDdmGZVkAGVwGwNJSFrbIPHsIhCDFABCStFmg2DsHgxDBiAA)

```ts
// -snip-
type Point = PartialBrand<RemoveIndex<[number, number]>, typeof pointId>;
//                        ^^^^^^^^^^^
```

[Playground link](https://www.typescriptlang.org/play?target=99&jsx=0&ssl=22&ssc=38&pln=21&pc=1#code/PTAEAsBdIBwZwFwjpAhgYwNYHsBuBTAJwDMAbbAdwDp1sBbYVYANmYCYBWN5j5gWABQkAJ4x8oAEr46efAEkAdgBN8ADwA8AFQB8oALygA3oNCgA2qADSoAJYLQmfMOzFQm0KjgnToFITsA5qBqkPjKcFbepgD8oAr4BIRRoAhxAK50AEZEwaqh4ZECPqCx8YnJqXDCWdikuflKEZbJpQlEFYWmALopbmaWXQDcggC+wwKCKuikqITitAootkqpaQo2AI5p4lU1pOMiYqBySlq6Boagc6hK2AqkwuY2Sl2p7mOCgofiAEKEqMotAAaUAAVXOblAADJjqdwQdROIAAqzSA2VCkP4A06aEHg-SQmEowhojHqE7qcHaQagECgACCpDg2AcCko9k8oAAYjNcNhCFRPgIpjM5qAFksYNg7JATqt1lsdtVMrUEUckdKFJACcTSZj-oCpDICIoVBozAoMtlCCDLVkiF1tCDvi5QFKZSdqYI6cVfX7-aAAHrBkOhwNCiXa91agkAClUqTt1oJAAYQcJE1acgYUwBKVIamX6XRmVTpoYRu5LG5KOOoAuayAgzINmUg9CtmMGaOQWO5-OgQtd3TGIo+dBmFM9AyoSc9ADUoEyc-GxQnAEZpx4zJvQIvl5vVz45pA0oR7Ohxh8JgJIx4CT3Y2nQOvc+M75kH43Y2wQQBmN9BDvdACRrWNUGbQCbwlWp8CocgAljdAoKAucCQAFnfHctw4LC2C3ZgaTpABlcBsDSUha2yDx7CIQh+QAQkrRZYPg7BEOQwYgA)
